src/main/java/de/tum/in/www1/artemis/service/user/UserService.java

```diff
@@ -236,10 +240,26 @@ public User registerUser(UserDTO userDTO, String password) {
         Set<Authority> authorities = new HashSet<>();
         authorityRepository.findById(STUDENT.getAuthority()).ifPresent(authorities::add);
         newUser.setAuthorities(authorities);
-        saveUser(newUser);
+
         // we need to save first so that the user can be found in the database in the subsequent method
-        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> vcsUserManagementService.createVcsUser(newUser));
-        optionalCIUserManagementService.ifPresent(ciUserManagementService -> ciUserManagementService.createUser(newUser));
+        saveUser(newUser);
+
+        optionalVcsUserManagementService.ifPresent(vcsUserManagementService -> {
+            if (!vcsUserManagementService.canCreateVcsUser(newUser)) {
+                // Delete the user that was previously saved and throw
+                deleteUser(newUser);
+                throw new VCSAccountExistsException();
+            }
+        });
+
+        optionalCIUserManagementService.ifPresent(ciUserManagementService -> {
+            if (!ciUserManagementService.canCreateUser(newUser)) {
+                // Delete the user that was previously saved and throw
+                deleteUser(newUser);
+                throw new CIAccountExistsException();
```

Comment:
provide more details in the exception for better error handling (e.g. the username)